domains/client/domain-executor/src/xdm_verifier.rs

```diff
@@ -99,18 +108,24 @@ where
     PBlock: BlockT,
     NumberFor<PBlock>: From<NumberFor<SBlock>>,
     PBlock::Hash: From<SBlock::Hash>,
+    SExecutorDispatch: NativeExecutionDispatch + 'static,
 {
-    let system_domain_runtime = system_domain_client.runtime_api();
-    let best_hash = system_domain_client.info().best_hash;
-    if let Ok(Some(state_roots)) =
-        system_domain_runtime.extract_xdm_proof_state_roots(best_hash, extrinsic)
-    {
-        // TODO: Can't above variable be reused here?
+    let primary_api = primary_chain_client.runtime_api();
+    let primary_best_hash = primary_chain_client.info().best_hash;
+    let system_domain_wasm = primary_api.system_domain_wasm_bundle(primary_best_hash)?;
+
+    if let Ok(state_roots) = call_function_on_wasm_stateless::<
+        _,
+        ExtractedStateRootsFromProof<NumberFor<SBlock>, SBlock::Hash, SBlock::Hash>,
+        SExecutorDispatch,
+    >(
+        system_domain_wasm.as_ref(),
+        "MessengerApi_extract_xdm_proof_state_roots",
```

Comment:
Not a fan on such string arguments, but I don't necessarily have a better alternative. The concern is that these functions are easy to break during refactoring and such with no compiler help to detect that.

Reply:
I agree on not able to ensure compatibility but I dont see any alternative without bringing in unneeded logic to primary runtime to achieve something similar. I'm still thinking on how to ensure this wont break on refactor. No strong solutions yet.

As for string arguments, TBH, I kinda like it. Though you can make it nicer with some macros like substrate does.